### ISSUE TITLE
Importing User List or Creating a Custom Source?

### ISSUE THREAD

**Issue by ajquick** [goauthentik/authentik#6796]
Hello all,

I just downloaded Authentik and got through the basic setup. I have a decent understanding how Authentik works as an identity provider, authentication server and has the ability federate / social login as well as handle the data keeping itself. It seems like it is so extremely capable when dealing with login systems that already use LDAP, SAML or OAuth... etc.

What I am trying to wrap my head around is how do I get my existing user database INTO Authentik if it doesn't use one of the existing authorization schemes?

Under normal circumstances, I would already have an LDAP or OAuth provider that Authentik could federate to. However, I am using a closed off login ecosystem from [phpBB](https://phpbb.com). They do not support any kind of authorization services against their logins. Would I need to create a custom phpBB source? Is there a way to extend Authentik to create a custom source?

I have a MySQL database with a table of users with usernames, emails, uid and hashed password. I have considered writing a middleware wrapper that uses SAML or otherwise to allow Authentik to federate the login to the database. I would however, much rather just import all the logins at once.

Does Authentik have an easy way to import, say from a CSV list of users?

Could Authentik have a barebones feature where it could connect to an external database and perhaps map various fields from that database to the internal user database?

Could it be as simple as editing the Postgres database and importing users manually?

Does Authentik support alternative hashing methods? I imagine some of my oldest users in the database still have md5 hashed passwords, or some variation phpBB's hash method. I don't think I could import those directly. Is there a way to specify what hashing method was used on the password and then update it to something more secure upon the next login?

Just trying to figure out the best way to go about this without having everyone reset their passwords.

**Comment by DuvelCorp** [comment]
Hi

I am glad to see this fresh question popping up because I am exactly  in the same situation.

I've got a Family and Friends server used by more or less 30 users that have an account in Jellyfin, (+ Jellyseer and  Organizr synced from Jellyfin), and Komga.
I have set up authenthik with ldap and all is working fine. I can connect on Jellyfin from Authentik with a testuser I created both sides.
Now my brains bugs and I am totally stuck for the next step.  This is not a technical issue, this is a functional issue: 
I just don't know what to do with my existing user base. Obviously they should be created in Authentik. But HOW??? 
I searched the web extensively and it is nowhere. I have the impression that most selhosted people migrating to Authentik just have their own account and eventually their wife or a friend but that's all.

The only approach I see now is to  create each user manually in authenik, with a temp password and invite them somehow to login with a token and force them to change their password. I guess I should also change the Auth provider for each of them in their Jellyfin settings. 
All that seems to be a very cumbersome operation with a medium user base like I have, and seems pretty impossible to do with a large user base, so am I missing something???

Cheers!

**Comment by ajquick** [comment]
I'm glad I'm not alone!

I do think the intention is to start fresh and create new accounts OR connect an existing LDAP or similar to interface and federate your users over to Authentik.

**Comment by DuvelCorp** [comment]
To  bad we haven't received any answer and beginning of solutions or proposals that might put us on the right direction

**Comment by ajquick** [comment]
@DuvelCorp 

I wanted to let you know what I am planning on doing to make this work.

I downloaded a library called [SimpleSAMLphp](https://github.com/simplesamlphp/simplesamlphp). While I was initially hesitant to use the library, as many on StackOverflow seem to believe it is a fairly poorly written piece of software, it appears to be fairly robust currently.

I am following the documentation:

* [Installation](https://simplesamlphp.org/docs/stable/simplesamlphp-install.html)
* [Using as an IdP](https://simplesamlphp.org/docs/stable/simplesamlphp-idp.html)
* [Custom authentication](https://simplesamlphp.org/docs/2.0/simplesamlphp-customauth.html)

Basically I should be able to write my own module that connects to my user database and checks the login credentials and returns the information I want it to return back to Authentik.

I should have it setup eventually to use Authentik as the SP, which will federate the login request to my IdP database. Then my existing software should use ldap or similar to check Authentik for the login. I would like to get SSO going after that.

**Comment by ajquick** [comment]
I'm back at square one.

I thought Authentik had a way to transfer the username and password to the third party and do a migrate on authentication process. Apparently that is something FusionAuth does as a paid feature.

I can't see the point in using Authentik if they still need to login somewhere else. I was hoping this would be more like a free version of Auth0 or FusionAuth.

**Comment by DrZoidberg09** [comment]
Hi there,

Same topic for me. It would really cool to have a build in function to import users via CSV/XLS files.

The best way that comes to my mind (not tested yet) would be to write a Python script using the Authentik API to upload users from a CSV/XLS.

**Comment by BeryJu** [comment]
The API endpoints https://docs.goauthentik.io/developer-docs/api/reference/core-users-create and https://docs.goauthentik.io/developer-docs/api/reference/core-users-set-password-create can be used to create users from a script. We've decided to not add an example script as most of the script would vary depending on what source users are being read from.

**Comment by fritz-c** [comment]
I've been theorizing about how to do this - import my database users gracefully into Authentik's DB, using the existing password hashes as-is from the old database until they can be overwritten with whatever hash algorithm Authentik prefers.

I did some digging into the code, and it appears that the Password stage just [calls a module from Django to check existing password hashes](https://github.com/goauthentik/authentik/blob/cc0659168d56a31c30ec05b254fac7b0b208779d/authentik/core/auth.py#L21). From there, Django has a bit of flexible logic around checking the hash, in that [it will look for a prefix on the hash to attempt to identify the hash algorithm](https://github.com/django/django/blob/5a1cae3a5675c5733daf5949759476d65aa0e636/django/contrib/auth/hashers.py#L167-L186). If the prefix is one of the known hash algos but not the preferred one, we're in luck - after verifying with the original algorithm, it re-hashes with the preferred algorithm and saves it. Which hashing algorithms are available is dependent on the [`PASSWORD_HASHERS` setting](https://github.com/django/django/blob/5a1cae3a5675c5733daf5949759476d65aa0e636/django/conf/global_settings.py#L539-L545) in Django, the first of which is considered the preferred algorithm. [The hash algorithms themselves are found back in the same `hashers.py` file](https://github.com/django/django/blob/5a1cae3a5675c5733daf5949759476d65aa0e636/django/contrib/auth/hashers.py#L311-L688) that had the hash identification code. You might think at this point that we've reached the finish line - we just add the right prefix to our hashes, plug them into the Authentik users DB somehow, and be done. Unfortunately, though, the hash algorithms Django defines are a bit quirky, doing things like a sha256 hash on a password before hashing it with bcrypt, so it's unlikely your passwords from another system are compatible with the existing set (still, check the [available hash algorithm implementations](https://github.com/django/django/blob/5a1cae3a5675c5733daf5949759476d65aa0e636/django/contrib/auth/hashers.py#L311-L688) just to be sure) (EDIT: as @jbrinksmeier pointed out below, django does have a more standard bcrypt implementation available, it's just not in `PASSWORD_HASHERS`).

There is an apparent workaround though. It turns out we can modify that `PASSWORD_HASHERS` setting by [using the `/data/user_settings.py` file described in the docs](https://docs.goauthentik.io/docs/install-config/configuration/#custom-python-settings). With the ability to add that, we can plug in our own hash algorithms that generate a hash with identical logic to the old platform we're migrating from. We can even make ours the preferred hash (putting it in the front of the list) if we're confident in it. There is a [base class you build hash algorithms with](https://github.com/django/django/blob/5a1cae3a5675c5733daf5949759476d65aa0e636/django/contrib/auth/hashers.py#L204), and I think you could even define your algorithm in that same file, referring to it via `data.user_settings.MyGreatHasherClass` or the like inside the `PASSWORD_HASHERS` list. I'm a bit rusty on Python, though, so I could be wrong. If your hash algorithm has extra dependencies beyond those included with Authentik, I'm sure there's some workaround where you could mount your own directory containing your hash algo code and its own venv in the docker container.

With the ability to verify your old hashes in place, now just the user migration remains. The [`core_users_create`](https://docs.goauthentik.io/developer-docs/api/reference/core-users-create) API mentioned by @BeryJu would probably come most in handy here. You make a script that, for each user in your old DB, calls the `core_users_create` API with appropriate parameters, finds the newly created user record in the Authentik DB, and copies over the old password hash to the new DB, adding the algorithm prefix you defined in your custom hasher.

In theory, that's it 😅

**Comment by jbrinksmeier** [comment]
Thank you @fritz-c for your super helpful insights.
As our use-case involves bcrypt hashed passwords, we obviously faced your comment about the bcrypt-sha256 quirkyness, and I think its noteworthy here that django does provide a "normal" bcrypt hasher implementation. Its simply not in the `PASSWORD_HASHERS` list, and adding it via the `user_settings.py`:

````
PASSWORD_HASHERS = [
    "django.contrib.auth.hashers.PBKDF2PasswordHasher",
    "django.contrib.auth.hashers.PBKDF2SHA1PasswordHasher",
    "django.contrib.auth.hashers.Argon2PasswordHasher",
    "django.contrib.auth.hashers.BCryptSHA256PasswordHasher",
    "django.contrib.auth.hashers.ScryptPasswordHasher",
    "django.contrib.auth.hashers.BCryptPasswordHasher"
]
````
did the trick for us. We simply imported the passwords from our own user db, added a `bcrypt$` prefix to all strings, and were good to go.
